test/e2e/upgrade_test.go

```diff
@@ -413,5 +413,10 @@ var _ = Describe("Upgrade", Label("optional", "upgrade"), func() {
 
 	AfterEach(func() {
 		cleanup(ctx, upgradeClusterProxy, namespace, cancelWatches, e2eConfig.GetIntervals("default", "wait-namespace-deleted")...)
+		// The BMO/Ironic namespace is deleted after each test, but we need to ensure that it actually gets deleted.
+		WaitForNamespaceDeleted(ctx, WaitForNamespaceDeletedInput{
```

Comment:
Or maybe even before.

Reply:
Ugh this is getting too ugly. The problem is that the namespace is part of both the ironic and bmo kustomizations. It will be deleted when we delete either of them.

Reply:
Shouldn't it still be there unless all of them are deleted?

Anyway, IMO it's not really an issue, if the namespace has been deleted already when the `DeferCleanup()` is called, then the `WaitForNamespaceDeleted()` function should just succeed, right?

Reply:
Yeah it's not an issue in that regard, it is just very ugly that we create and delete the namespace in both these.
The namespace definitely is deleted with both BMO and Ironic since both the [ironic](https://github.com/metal3-io/baremetal-operator/blob/34e12993f24970b3a23ae116f45f4db39d7ef80c/ironic-deployment/overlays/e2e/kustomization.yaml#L5) and [BMO](https://github.com/metal3-io/baremetal-operator/blob/34e12993f24970b3a23ae116f45f4db39d7ef80c/config/overlays/basic-auth_tls/kustomization.yaml#L5) overlays include the namespace. If you check the logs you will also see tons of errors about "no such object" for all of those that have already been deleted.
In practice, we always delete BMO when we delete Ironic and vice versa. It is not great and there is a chance that this can cause issues. For example, deleting BMO before all BMHs are gone will make them stuck.

Reply:
Hmm. I don't think the namespace would really be cleaned when we delete BMO, if ironic still lives in it, but I'm not 100% sure. 
Anyway, do you have a suggested solution here? Maybe we should just remove the namespace from both the overlays, and explicitly create the namespace before installation?

Reply:
I am 100 % sure that the namespace is deleted when we delete BMO. It *also* deletes Ironic. The equivalent of `kubectl delete namespace baremetal-operator-system` is executed when we delete BMO, which will delete everything in the namespace first and then the namespace itself.

> Anyway, do you have a suggested solution here? Maybe we should just remove the namespace from both the overlays, and explicitly create the namespace before installation?

There are a few options:
1. always bundle BMO and Ironic so they are always installed as one unit (which is impractical)
2. give them separate namespaces
3. be more fine grained so we leave the namespace alone (this could include an operator that takes care of all operations)
4. handle the namespace separately, a bit annoying and would require changes in multiple places

> That's a valid concern, but I don't think either BMO or Ironic deletion will ever happen after BMH cleanup.

That may be true now and for these tests but not in general. We could easily have a test where we try deleting and re-creating Ironic in the middle of a test to see that it can recover. That would easily break if BMO was also deleted. (If CRDs are deleted then the BMHs also go.)

Reply:
I see. So you are mentioning this issue for general use as well, not just in BMO e2e tests? If that's the case, I have no idea, every option seems to have some side effect.